### PR TITLE
add SCC for openshift

### PR DESCRIPTION
**Pull request by jshaughn**

### CODE REVIEW

File: hack/istio/install-bookinfo-demo.sh [hunk 1]
```diff
@@ -140,6 +140,18 @@ else
   exit 1
 fi
 
+IS_OPENSHIFT="false"
+IS_MAISTRA="false"
+if [[ "${CLIENT_EXE}" = *"oc" ]]; then
+  IS_OPENSHIFT="true"
+  if [[ $([ "$(oc get crd | grep servicemesh | wc -l)" -gt "0" ] && echo "true") ]]; then
+    IS_MAISTRA="true"
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:148]: that's alot of extra stuff you don't need (that `&& echo "true"` is superfluous the way you have it). If you really want the value to be false as opposed to empty string when not maistra, then I suggest:
```suggestion
  IS_MAISTRA=$([ "$(oc get crd | grep servicemesh | wc -l)" -gt "0" ] && echo "true" || echo "false")
```

File: hack/istio/install-bookinfo-demo.sh [hunk 2]
```diff
@@ -211,15 +226,34 @@ $CLIENT_EXE get services -n ${NAMESPACE}
 $CLIENT_EXE get pods -n ${NAMESPACE}
 
 # If OpenShift, we need to do some additional things
-if [[ "$CLIENT_EXE" = *"oc" ]]; then
+if [[ "${IS_OPENSHIFT}" = "true" ]]; then
   $CLIENT_EXE expose svc/productpage -n ${NAMESPACE}
   $CLIENT_EXE expose svc/istio-ingressgateway --port http2 -n ${ISTIO_NAMESPACE}
-  cat <<NAD | $CLIENT_EXE -n ${NAMESPACE} create -f -
+  if [[ "${IS_MAISTRA}" == "true" ]]; then
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:232]: You mean if NOT maistra - Maistra will create the NAD for us. If we are NOT in Maistra, then we have to do this ourselves.
```suggestion
  if [[ "${IS_MAISTRA}" != "true" ]]; then
```

File: hack/istio/install-bookinfo-demo.sh [hunk 3]
```diff
@@ -211,15 +226,34 @@ $CLIENT_EXE get services -n ${NAMESPACE}
...
+seLinuxContext:
+  type: RunAsAny
+supplementalGroups:
+  type: RunAsAny
+users:
+- "system:serviceaccount:${NAMESPACE}:bookinfo-details"
+- "system:serviceaccount:${NAMESPACE}:bookinfo-productpage"
+- "system:serviceaccount:${NAMESPACE}:bookinfo-ratings"
+- "system:serviceaccount:${NAMESPACE}:bookinfo-ratings=v2"
+- "system:serviceaccount:${NAMESPACE}:bookinfo-reviews"
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:255]: this is fine. If they don't exist, they should just be ignored IIRC. But what about things like mongo and mysql? Do those have SAs, too?

**Review by jmazzitelli — CHANGES_REQUESTED** [review]

File: hack/istio/install-bookinfo-demo.sh [hunk 4]
```diff
@@ -140,6 +140,18 @@ else
   exit 1
 fi
 
+IS_OPENSHIFT="false"
+IS_MAISTRA="false"
+if [[ "${CLIENT_EXE}" = *"oc" ]]; then
+  IS_OPENSHIFT="true"
+  if [[ $([ "$(oc get crd | grep servicemesh | wc -l)" -gt "0" ] && echo "true") ]]; then
+    IS_MAISTRA="true"
+  fi
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:149]: If you take my recommendation above, remove this `fi` - I forgot to include its removal in my diff above.
```suggestion
```

File: hack/istio/install-bookinfo-demo.sh (hunk 3 above)
Comment (jshaughn) [reply hack/istio/install-bookinfo-demo.sh:254]: I didn't see them.  Should I add default as well, just to cover the bases?

Comment (jmazzitelli) [reply hack/istio/install-bookinfo-demo.sh:253]: I wouldn't add "default" unless you know you need it.

File: hack/istio/install-bookinfo-demo.sh [hunk 5]
```diff
@@ -161,7 +174,7 @@ if [ "${DELETE_BOOKINFO}" == "true" ]; then
 fi
 
 # If OpenShift, we need to do some additional things
-if [[ "$CLIENT_EXE" = *"oc" ]]; then
+if [[ "${IS_OPENSHIFT}" == "true" ]]; then
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:177]: use single-brackets with `==` - no need anymore for the double-brackets (which was needed due to the `= *` feature):
```suggestion
if [ "${IS_OPENSHIFT}" == "true" ]; then
```

File: hack/istio/install-bookinfo-demo.sh [hunk 6]
```diff
@@ -211,15 +224,35 @@ $CLIENT_EXE get services -n ${NAMESPACE}
 $CLIENT_EXE get pods -n ${NAMESPACE}
 
 # If OpenShift, we need to do some additional things
-if [[ "$CLIENT_EXE" = *"oc" ]]; then
+if [[ "${IS_OPENSHIFT}" = "true" ]]; then
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:227]: use single-brackets and double-equals:
```suggestion
if [ "${IS_OPENSHIFT}" == "true" ]; then
```

File: hack/istio/install-bookinfo-demo.sh [hunk 7]
```diff
@@ -211,15 +224,35 @@ $CLIENT_EXE get services -n ${NAMESPACE}
...
   $CLIENT_EXE expose svc/productpage -n ${NAMESPACE}
   $CLIENT_EXE expose svc/istio-ingressgateway --port http2 -n ${ISTIO_NAMESPACE}
-  cat <<NAD | $CLIENT_EXE -n ${NAMESPACE} create -f -
+  if [[ "${IS_MAISTRA}" != "true" ]]; then
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:230]:
```suggestion
  if [ "${IS_MAISTRA}" != "true" ]; then
```

File: hack/istio/install-bookinfo-demo.sh [hunk 8]
```diff
@@ -150,8 +160,11 @@ fi
 
 if [ "${DELETE_BOOKINFO}" == "true" ]; then
   echo "====== UNINSTALLING ANY EXISTING BOOKINFO DEMO ====="
-  if [[ "$CLIENT_EXE" = *"oc" ]]; then
-    $CLIENT_EXE delete network-attachment-definition istio-cni -n ${NAMESPACE}
+  if [[ "${IS_OPENSHIFT}" = "true" ]]; then
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:163]: double-equals, single brackets
```suggestion
  if [ "${IS_OPENSHIFT}" == "true" ]; then
```

File: hack/istio/install-bookinfo-demo.sh [hunk 9]
```diff
@@ -150,8 +160,11 @@ fi
...
+    if [[ "${IS_MAISTRA}" != "true" ]]; then
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:164]:
```suggestion
    if [ "${IS_MAISTRA}" != "true" ]; then
```

File: hack/istio/install-fraud-detection-demo.sh [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+  IS_OPENSHIFT="true"
+  IS_MAISTRA=$([ "$(oc get crd | grep servicemesh | wc -l)" -gt "0" ] && echo "true" || echo "false")
+fi
+
+echo "IS_OPENSHIFT=${IS_OPENSHIFT}"
+echo "IS_MAISTRA=${IS_MAISTRA}"
+
+if [ "${DELETE_DEMO}" == "true" ]; then
+  if [[ "${IS_OPENSHIFT}" = "true" ]]; then
+    if [[ "${IS_MAISTRA}" != "true" ]]; then
```
Comment (jmazzitelli) [inline hack/istio/install-fraud-detection-demo.sh:67]:
```suggestion
  if [ "${IS_OPENSHIFT}" == "true" ]; then
    if [ "${IS_MAISTRA}" != "true" ]; then
```

File: hack/istio/install-fraud-detection-demo.sh [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    fi
+    $CLIENT_EXE delete security-context-constraints fraud-detection-scc -n ${NAMESPACE}
+  fi
+  ${CLIENT_EXE} delete namespace ${NAMESPACE}
+  exit 0
+fi
+
+# Create and prepare the demo namespaces
+
+if [[ "${IS_OPENSHIFT}" == "true" ]]; then
```
Comment (jmazzitelli) [inline hack/istio/install-fraud-detection-demo.sh:78]:
```suggestion
if [ "${IS_OPENSHIFT}" == "true" ]; then
```

File: hack/istio/install-fraud-detection-demo.sh [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+${CLIENT_EXE} apply -f <(curl -L "${SOURCE}/fraud-detection/accounts.yaml") -n ${NAMESPACE}
+${CLIENT_EXE} apply -f <(curl -L "${SOURCE}/fraud-detection/cards.yaml") -n ${NAMESPACE}
+${CLIENT_EXE} apply -f <(curl -L "${SOURCE}/fraud-detection/bank.yaml") -n ${NAMESPACE}
+${CLIENT_EXE} apply -f <(curl -L "${SOURCE}/fraud-detection/policies.yaml") -n ${NAMESPACE}
+${CLIENT_EXE} apply -f <(curl -L "${SOURCE}/fraud-detection/claims.yaml") -n ${NAMESPACE}
+${CLIENT_EXE} apply -f <(curl -L "${SOURCE}/fraud-detection/insurance.yaml") -n ${NAMESPACE}
+${CLIENT_EXE} apply -f <(curl -L "${SOURCE}/fraud-detection/fraud.yaml") -n ${NAMESPACE}
+
+if [[ "${IS_OPENSHIFT}" = "true" ]]; then
+  if [[ "${IS_MAISTRA}" != "true" ]]; then
```
Comment (jmazzitelli) [inline hack/istio/install-fraud-detection-demo.sh:99]:
```suggestion
if [ "${IS_OPENSHIFT}" == "true" ]; then
  if [ "${IS_MAISTRA}" != "true" ]; then
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
use if syntax consistently - double-brackets have special meaning and we only needed it with the `= *` syntax. Use single-brackets, double-equals for consistency and better compatibility with most bash impls

File: hack/istio/install-bookinfo-demo.sh (hunk 3 above)
Comment (jshaughn) [reply hack/istio/install-bookinfo-demo.sh:254]: I see services in all the demos that don't have a serviceaccount, so I think we may need it

File: hack/istio/install-fraud-detection-demo.sh [hunk 13]
```diff
@@ -62,6 +62,7 @@ fi
 echo "IS_OPENSHIFT=${IS_OPENSHIFT}"
 echo "IS_MAISTRA=${IS_MAISTRA}"
 
+# If we are to delete, remove everything and exit immediately after
 if [ "${DELETE_DEMO}" == "true" ]; then
```
Comment (jmazzitelli) [inline hack/istio/install-fraud-detection-demo.sh:66]: is it me or are you missing the `fi` to close this if-stmt?

Comment (jshaughn) [reply hack/istio/install-fraud-detection-demo.sh:66]: that's a comment

Comment (jshaughn) [reply hack/istio/install-fraud-detection-demo.sh:66]: I dunno, maybe you need to refresh

**Comment by jmazzitelli** [comment]
I was going to run these, but then noticed it won't work for me as-is. I'm using Maistra and the auto-injection features of the script assumes upstream Istio (it will label the namespace but will not add the annotation to the Deployment - Maistra does not support the namespace-wide injection label).

I will attempt to fix and submit a PR to this PR if I get it to work. All three scripts need to change.

File: hack/istio/install-bookinfo-demo.sh [hunk 14]
```diff
@@ -148,10 +158,14 @@ if [ "${GATEWAY_YAML}" == "" ]; then
...
+# If we are to delete, remove everything and exit immediately after
 if [ "${DELETE_BOOKINFO}" == "true" ]; then
   echo "====== UNINSTALLING ANY EXISTING BOOKINFO DEMO ====="
-  if [[ "$CLIENT_EXE" = *"oc" ]]; then
-    $CLIENT_EXE delete network-attachment-definition istio-cni -n ${NAMESPACE}
+  if [ "${IS_OPENSHIFT}" == "true" ]; then
+    if [ "${IS_MAISTRA}" != "true" ]; then
+      $CLIENT_EXE delete network-attachment-definition istio-cni -n ${NAMESPACE}
+    fi
+    $CLIENT_EXE delete security-context-constraints bookinfo-scc -n ${NAMESPACE}
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:168]: this is wrong, you get `error: the server doesn't have a resource type "security-context-constraints"`

change to `scc`
```suggestion
    $CLIENT_EXE delete scc bookinfo-scc -n ${NAMESPACE}
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]

File: hack/istio/install-bookinfo-demo.sh [hunk 15]
```diff
@@ -211,20 +225,41 @@ $CLIENT_EXE get services -n ${NAMESPACE}
...
-  cat <<NAD | $CLIENT_EXE -n ${NAMESPACE} create -f -
+  if [ "${IS_MAISTRA}" != "true" ]; then
+    cat <<NAD | $CLIENT_EXE -n ${NAMESPACE} create -f -
 apiVersion: "k8s.cni.cncf.io/v1"
 kind: NetworkAttachmentDefinition
 metadata:
   name: istio-cni
 NAD
+  fi  
+  cat <<SCC | $CLIENT_EXE -n ${NAMESPACE} create -f -
```
Comment (jmazzitelli) [inline hack/istio/install-bookinfo-demo.sh:239]:
```suggestion
  cat <<SCC | $CLIENT_EXE create -f -
```

File: hack/istio/install-error-rates-demo.sh [hunk 16]
```diff
@@ -54,12 +54,26 @@ echo NAMESPACE_ALPHA=${NAMESPACE_ALPHA}
...
   echo "Deleting Error Rates Demo (the envoy filters, if previously created, will remain)"
-  if [ "${CLIENT_EXE}" == "oc" ]; then
-    ${CLIENT_EXE} delete network-attachment-definition istio-cni -n ${NAMESPACE_ALPHA}
-    ${CLIENT_EXE} delete network-attachment-definition istio-cni -n ${NAMESPACE_BETA}
+  if [ "${IS_OPENSHIFT}" == "true" ]; then
+    if [ "${IS_MAISTRA}" != "true" ]; then
+      $CLIENT_EXE delete network-attachment-definition istio-cni -n ${NAMESPACE_ALPHA}
+      $CLIENT_EXE delete network-attachment-definition istio-cni -n ${NAMESPACE_BETA}
+    fi
+    $CLIENT_EXE delete security-context-constraints error-rates-scc
```
Comment (jmazzitelli) [inline hack/istio/install-error-rates-demo.sh:76]:
```suggestion
    $CLIENT_EXE delete scc error-rates-scc
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]

File: hack/istio/install-bookinfo-demo.sh (hunk 14 above)
Comment (jmazzitelli) [reply hack/istio/install-bookinfo-demo.sh]: I fixed this in my PR

**Comment by jmazzitelli** [comment]
with my PR included:
```
$ oc get pods -n alpha
NAME                        READY   STATUS    RESTARTS   AGE
a-client-78949db855-944xx   2/2     Running   0          7m41s
b-client-cb4d659d5-xc886    2/2     Running   0          7m41s
c-client-59c5899574-bfksc   2/2     Running   0          7m40s
d-client-854f55b6f9-pqkln   2/2     Running   0          7m40s
e-client-6d644677f5-qmgjp   2/2     Running   0          7m40s
f-client-dd8dbf757-d9glc    2/2     Running   0          7m40s
v-server-5946544cd6-zc4zs   2/2     Running   0          7m39s
w-server-5c4db874d8-74gvg   2/2     Running   0          7m39s
x-server-799c84f9d8-cd5td   2/2     Running   0          7m39s
y-server-744669d699-2bkks   2/2     Running   0          7m39s
z-server-798b4c4498-4fpwb   2/2     Running   0          7m38s
$ oc get pods -n beta
NAME                        READY   STATUS    RESTARTS   AGE
a-client-5bb6874875-74sl2   2/2     Running   0          7m39s
b-client-847b4bcdf7-vhr4p   2/2     Running   0          7m39s
c-client-ff696b888-ld2r4    2/2     Running   0          7m39s
d-client-6f8b48c57-bdj47    2/2     Running   0          7m39s
e-client-7c695fc7f9-g4pd4   2/2     Running   0          7m38s
f-client-66c98cf849-kd4xz   2/2     Running   0          7m38s
v-server-7b598dbd5b-kthmk   2/2     Running   0          7m38s
w-server-5c4db874d8-n4vbz   2/2     Running   0          7m37s
x-server-799c84f9d8-h2hmt   2/2     Running   0          7m37s
y-server-744669d699-4dzkb   2/2     Running   0          7m37s
z-server-798b4c4498-p9cjp   2/2     Running   0          7m37s
$ oc get pods -n fraud-detection 
NAME                           READY   STATUS    RESTARTS   AGE
accounts-v1-54cb8f697c-2k6gk   2/2     Running   0          7m24s
bank-v1-76f584c854-s9lvp       2/2     Running   0          7m23s
cards-v1-55c4667fff-vvd4b      2/2     Running   0          7m23s
claims-v1-6c5f9d4c79-wqfqm     2/2     Running   0          7m23s
fraud-v1-78b5f5cfd4-bgnww      2/2     Running   0          7m23s
insurance-v1-bb4d7d986-9b588   2/2     Running   0          7m22s
policies-v1-887589db9-n5drt    2/2     Running   0          7m22s
$ oc get pods -n travel-agency
NAME                             READY   STATUS    RESTARTS   AGE
cars-v1-6b88556757-rjsb7         2/2     Running   0          6m55s
discounts-v1-5bdd64fdd-sd8p6     2/2     Running   0          6m54s
flights-v1-754d44d4f9-47lqm      2/2     Running   0          6m53s
hotels-v1-6dc9c978d5-2cl8r       2/2     Running   0          6m54s
insurances-v1-7c4b5749f8-sjg9r   2/2     Running   0          6m53s
mysqldb-v1-7bd654bf74-jm7c2      2/2     Running   0          6m54s
travels-v1-76f7467bb8-vlxjj      2/2     Running   0          6m53s
$ oc get pods -n travel-control
NAME                       READY   STATUS    RESTARTS   AGE
control-5fcd7bb48d-cmxww   1/1     Running   0          6m55s
$ oc get pods -n travel-portal
NAME                       READY   STATUS    RESTARTS   AGE
travels-7fd7c55466-g9wwf   2/2     Running   2          7m5s
viaggi-666dbc746b-mcn24    2/2     Running   2          7m5s
voyages-566d4b4d49-qwxhj   2/2     Running   2          7m5s
```
not sure why control namespace doesn't have its pod injected - the annotation exists.

graph looks kinda sane
image

**Comment by jmazzitelli** [comment]
now passing in `-d true` (deleting) to all the scripts:
```
$ ./install-error-rates-demo.sh -d true
Will deploy Error Rates Demo using these settings:
CLIENT_EXE=oc
DELETE_DEMO=true
ENABLE_INJECTION=true
ISTIO_NAMESPACE=istio-system
NAMESPACE_ALPHA=alpha
NAMESPACE_BETA=beta
SOURCE=https://raw.githubusercontent.com/kiali/demos/master
IS_OPENSHIFT=true
IS_MAISTRA=true
Deleting Error Rates Demo (the envoy filters, if previously created, will remain)
servicemeshmember.maistra.io "default" deleted
servicemeshmember.maistra.io "default" deleted
securitycontextconstraints.security.openshift.io "error-rates-scc" deleted
namespace "alpha" deleted
namespace "beta" deleted
$ ./install-fraud-detection-demo.sh -d true
Will deploy Error Rates Demo using these settings:
CLIENT_EXE=oc
DELETE_DEMO=true
ENABLE_INJECTION=true
ISTIO_NAMESPACE=istio-system
NAMESPACE=fraud-detection
SOURCE=https://raw.githubusercontent.com/kiali/demos/master
IS_OPENSHIFT=true
IS_MAISTRA=true
servicemeshmember.maistra.io "default" deleted
securitycontextconstraints.security.openshift.io "fraud-detection-scc" deleted
namespace "fraud-detection" deleted
$ ./install-travel-agency-demo.sh -d true
Will deploy Travel Agency using these settings:
CLIENT_EXE=oc
DELETE_DEMO=true
ENABLE_INJECTION=true
ENABLE_OPERATION_METRICS=false
ISTIO_NAMESPACE=istio-system
NAMESPACE_AGENCY=travel-agency
NAMESPACE_CONTROL=travel-control
NAMESPACE_PORTAL=travel-portal
SOURCE=https://raw.githubusercontent.com/kiali/demos/master
IS_OPENSHIFT=true
IS_MAISTRA=true
Deleting Travel Agency Demo (the envoy filters, if previously created, will remain)
servicemeshmember.maistra.io "default" deleted
servicemeshmember.maistra.io "default" deleted
servicemeshmember.maistra.io "default" deleted
securitycontextconstraints.security.openshift.io "travel-scc" deleted
namespace "travel-agency" deleted
namespace "travel-portal" deleted
namespace "travel-control" deleted
```

**Comment by jmazzitelli** [comment]
Ran travel agency demo again, and now I see sidecar - I have a feeling that was a maistra issue:
```
$ oc get pods -n travel-control 
NAME                       READY   STATUS    RESTARTS   AGE
control-5fcd7bb48d-mghqn   2/2     Running   0          23s
```

File: hack/istio/install-error-rates-demo.sh (hunk 16 above)
Comment (jmazzitelli) [reply hack/istio/install-error-rates-demo.sh:76]: fixed in my PR here https://github.com/jshaughn/kiali/pull/5/files#diff-be5229a9ec29e499939f17823afb8a72b1dfdea8cb386d7b5ef1245f5fce1a9dR113
nothing to do other than just merge my PR.
you should rebase your PR first.

**Comment by jshaughn** [comment]
@jmazzitelli OK, merged in your maistra additions for SMMR and I think these are now ready for merge...

**Review by jmazzitelli — APPROVED** [review]
lgtm

**Pull request merged**